src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -68,6 +74,34 @@ public void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
                 source.Cancel();
             }
+
+            ClearDiagnostics(workspace);
+
+            AssertCleanup();
+        }
+
+        private void ClearDiagnostics(Workspace workspace)
+        {
+            var analyzer = _provider.CreateIncrementalAnalyzer(workspace);
+            var args = _service.GetDiagnosticsUpdatedEventArgs(workspace, projectId: null, documentId: null, cancellationToken: CancellationToken.None);
+
+            foreach (var documentId in args.Select(a => a.DocumentId).Distinct())
+            {
+                analyzer.RemoveDocument(documentId);
```

Comment:
Here's part of the API for IIncrementalAnalyzer:

```c#
        void RemoveDocument(DocumentId documentId);
        void RemoveProject(ProjectId projectId);
```

It isn't clear at all why this code only calls RemoveDocument and not RemoveProject.  Furthermore, there's no indication from this API that calling RemoveDocument is the appropriate way to clear out cached data.  I'd far prefer the IIncrementalAnalyzer API get a method like ```ClearCachedData(Workspace)``` which would much more cleanly identify what needs to happen.  Each analyzer implementation could then appropriately dump it's information.

This is fairly important as i don't see any reason why .RemoveDocument will do the right thing across all our analyzer implementations (both now and in the future).